server/routes/experiments/biomed_nl/api.py

```diff
@@ -21,76 +22,145 @@
-                             start_entity,
-                             start_dcids,
-                             gemini_client,
-                             gemini_model_str=GEMINI_PRO)
+  traversed_entity_info = {}
+  try:
+    path_finder.find_start_and_traversal_type()
 
-    path_finder.find_paths()
-    # TODO: Add error handling before constructing cache.
+    if path_finder.traversal_type == PathFinder.TraversalTypes.OVERVIEW:
+      # Skip traversal and fetch data for the entities
+      path_finder.path_store.selected_paths = {
+          dcid: {} for dcid in path_finder.start_dcids
+      }
+    else:
+      path_finder.find_paths()
 
     traversed_entity_info = path_finder.get_traversed_entity_info()
-    # TODO: add error handling.
 
-    final_prompt = utils.FINAL_PROMPT.format(
-        sentence=query, json_str=utils.format_dict(traversed_entity_info))
+  except Exception as e:
+    logging.error(f'[biomed_nl]: {e}', exc_info=True)
+    if path_finder.start_dcids:
+      response[
+          'answer'] = f'{path_finder.start_entity_name}: {", ".join(path_finder.start_dcids)}'
+    else:
+      response[
+          'answer'] = 'Error finding entities from the query in the knowledge graph.'
+    return response
+
+  try:
+    response['debug'] += '\n' + utils.format_dict(
+        path_finder.path_store.get_paths_from_start(only_selected_paths=True))
+
+    # TODO: add error handling.
+    final_prompt = utils.FINAL_RESPONSE_PROMPT.format(
+        QUERY=query, ENTITY_INFO=utils.format_dict(traversed_entity_info))
 
     input_tokens = gemini_client.models.count_tokens(
         contents=final_prompt, model=GEMINI_PRO).total_tokens
-    gemini_response = None
-    if input_tokens < GEMINI_PRO_TOKEN_LIMIT:
+
+    should_include_fallback = False
+    if input_tokens < GEMINI_PRO_TOKEN_LIMIT * 0.75:
       gemini_response = gemini_client.models.generate_content(
-          model=GEMINI_PRO, contents=final_prompt)
+          model=GEMINI_PRO,
+          contents=final_prompt,
+          config={
+              'response_mime_type': 'application/json',
+              'response_schema': FinalAnswerResponse,
+          })
+      final_response = FinalAnswerResponse(**json.loads(gemini_response.text))
+      response['answer'] = final_response.answer
+      response['footnotes'] = '\n'.join(
+          [ref.model_dump_json(indent=2) for ref in final_response.references])
     else:
-      # Todo add log message
-      response['debug'] += '\nFetched data too large for Gemini'
-      return response
-
-    response['answer'] = gemini_response.text
-    response['debug'] += '\n' + utils.format_dict(
-        path_finder.path_store.get_paths_from_start(only_selected_paths=True))
+      should_include_fallback = True
 
+    # TODO: return fallback for traversal error + timeouts and uncertain final responses
+    if should_include_fallback:
+      _get_fallback_response(query, response, path_finder,
```

Comment:
you would want to either set this to the response or return this right?

Reply:
I updated the function name to append, so that we append this to the response.